### PR TITLE
[AIRFLOW-793] Enable compressed loading in S3ToHiveTransfer

### PR DESCRIPTION
**Pull request by krishnabhupatiraju**
Dear Airflow Maintainers,

Please accept this PR that addresses the following issues:
- [AIRFLOW-793](https://issues.apache.org/jira/browse/AIRFLOW-793)

Testing Done:
- Added new unit tests for the S3ToHiveTransfer module

### CODE REVIEW

**Comment by krishnabhupatiraju** [comment]
@artwr @aoen  would you please take a look

**Comment by jhsenjaliya** [comment]
@krishnabhupatiraju , you should squash all commits and update commit msg and PR title according to the guidelines. ref other PRs if you need any reference. Thanks

**Comment by krishnabhupatiraju** [comment]
@jhsenjaliya thanks. made changes. sorry about the first time, should have read more closely.

File: airflow/operators/s3_to_hive_operator.py [hunk 1]
```diff
@@ -91,6 +99,7 @@ def __init__(
             wildcard_match=False,
             s3_conn_id='s3_default',
             hive_cli_conn_id='hive_cli_default',
+            file_format='txt',
```
Comment (aoen) [inline airflow/operators/s3_to_hive_operator.py]: How about decompress=True/False instead? this makes the field a lot more clear.

Comment (artwr) [reply airflow/operators/s3_to_hive_operator.py]: Or even better, decompress based on the file extension.

File: airflow/operators/s3_to_hive_operator.py [hunk 2]
```diff
@@ -105,28 +114,40 @@ def __init__(
...
-        self.hive = HiveCliHook(hive_cli_conn_id=self.hive_cli_conn_id)
+        # Downloading file from S3
         self.s3 = S3Hook(s3_conn_id=self.s3_conn_id)
+        self.hive = HiveCliHook(hive_cli_conn_id=self.hive_cli_conn_id)
         logging.info("Downloading S3 file")
+
         if self.wildcard_match:
             if not self.s3.check_for_wildcard_key(self.s3_key):
-                raise AirflowException("No key matches {0}".format(self.s3_key))
+                raise AirflowException("No key matches {0}".
```
Comment (artwr) [inline airflow/operators/s3_to_hive_operator.py]: The `.` needs to be on the line below.

File: airflow/operators/s3_to_hive_operator.py [hunk 3]
```diff
@@ -136,33 +157,105 @@ def execute(self, context):
...
+                if self.check_headers:
+                    logging.info("Matching file header against field_dict")
+                    header_list = self._get_top_row_as_list(fn_uncompressed)
+                    if not self._match_headers(header_list):
+                        raise AirflowException("Header check failed")
+
+                # Deleting top header row
+                logging.info("Removing header from file {0}".
+                             format(fn_uncompressed))
+                headless_file = \
```
Comment (artwr) [inline airflow/operators/s3_to_hive_operator.py]:
```python
headless_file = self._delete_top_row_and_compress(
    fn_uncompressed,
    self.file_format,
    tmp_dir)
```

or

```
headless_file = (
    self._delete_top_row_and_compress(fn_uncompressed,
                                      self.file_format,
                                      tmp_dir))
```

File: airflow/operators/s3_to_hive_operator.py [hunk 4]
```diff
@@ -136,33 +157,105 @@ def execute(self, context):
...
+                logging.info("Loading file without headers into Hive")
+                self.hive.load_file(headless_file,
+                                    self.hive_table,
+                                    field_dict=self.field_dict,
+                                    create=self.create,
+                                    partition=self.partition,
+                                    delimiter=self.delimiter,
+                                    recreate=self.recreate)
+
+    def _uncompress_file(self, input_file_name, file_format, dest_dir):
```
Comment (artwr) [inline airflow/operators/s3_to_hive_operator.py]: This looks useful and could probably be ported to `airflow.utils`

File: airflow/operators/s3_to_hive_operator.py [hunk 5]
```diff
@@ -136,33 +157,105 @@ def execute(self, context):
...
+
+    def _get_top_row_as_list(self, file_name):
+        with open(file_name, 'rt') as f:
+            header_line = f.readline().strip()
+            header_list = header_line.split(self.delimiter)
+            return header_list
+
+    def _match_headers(self, header_list):
+        if not header_list:
+            raise AirflowException("Unable to retreive header row from file")
```
Comment (artwr) [inline airflow/operators/s3_to_hive_operator.py]: retrieve

File: airflow/operators/s3_to_hive_operator.py [hunk 6]
```diff
@@ -136,33 +157,105 @@ def execute(self, context):
...
+        field_names = list(self.field_dict.keys())
```
Comment (artwr) [inline airflow/operators/s3_to_hive_operator.py]: `keys()` is usually already a list, no ?

**Review by artwr — CHANGES_REQUESTED** [review]
I agree with Dan about the API. Maybe use decompress as a boolean flag and use the name of the file to automagically select the decompressing algorithm.

Some minor comments on formatting.

Thanks for the unittests!

File: airflow/operators/s3_to_hive_operator.py (hunk 1 above)
Comment (krishnabhupatiraju) [reply airflow/operators/s3_to_hive_operator.py]: Added a new boolean parameter called input_compressed. deriving file extension from filename

File: airflow/operators/s3_to_hive_operator.py (hunk 2 above)
Comment (krishnabhupatiraju) [reply airflow/operators/s3_to_hive_operator.py]: Done

File: airflow/operators/s3_to_hive_operator.py (hunk 3 above)
Comment (krishnabhupatiraju) [reply airflow/operators/s3_to_hive_operator.py]: Done

File: airflow/operators/s3_to_hive_operator.py (hunk 5 above)
Comment (krishnabhupatiraju) [reply airflow/operators/s3_to_hive_operator.py]: Type corrected.

File: airflow/operators/s3_to_hive_operator.py (hunk 6 above)
Comment (krishnabhupatiraju) [reply airflow/operators/s3_to_hive_operator.py]: Yes. Removed the explicit cast.

File: airflow/operators/s3_to_hive_operator.py (hunk 4 above)
Comment (krishnabhupatiraju) [reply airflow/operators/s3_to_hive_operator.py]: Created it under airflow/utils/compression.py. Tests created under tests/utils

**Comment by krishnabhupatiraju** [comment]
@artwr @aoen PTAL. Made all suggested changes. 
- Replaced file_format param with input_compressed
- Moved uncompress_file to utils
- Added test cases for uncompress_file
- Removed explicit list cast on self.field_dict.keys()
- Added all mentioned formatting changes

LMK if this looks ok. I will squash the comments and rebase the branch.

File: airflow/operators/s3_to_hive_operator.py [hunk 7]
```diff
@@ -105,28 +114,41 @@ def __init__(
...
+                raise AirflowException("No key matches {0}"
+                                       .format(self.s3_key))
             s3_key_object = self.s3.get_wildcard_key(self.s3_key)
         else:
             if not self.s3.check_for_key(self.s3_key):
                 raise AirflowException(
                     "The key {0} does not exists".format(self.s3_key))
             s3_key_object = self.s3.get_key(self.s3_key)
-        with NamedTemporaryFile("w") as f:
+        root, file_ext = os.path.splitext(s3_key_object.key)
```
Comment (artwr) [inline airflow/operators/s3_to_hive_operator.py:140]: nit: you might want to use `file_ext.lower()` in the rest of the code. I can totally see uppercase keys happening.

**Review by artwr — APPROVED** [review]
A minor nit, but the change set looks good to me now. Thanks for providing unit tests and taking the time to refactor!

File: airflow/operators/s3_to_hive_operator.py (hunk 7 above)
Comment (krishnabhupatiraju) [reply airflow/operators/s3_to_hive_operator.py:140]: Good catch. I will change that.

**Review by krishnabhupatiraju — COMMENTED** [review]
Made changes to ensure any case file extensions are handled

**Comment by artwr** [comment]
LGTM +1

**Comment by jlowin** [comment]
@artwr small issue -- this PR leads to an attempt to import `hive_metastore` any time unit tests are run, even if these aren't the unit tests being run. `hive_metastore` is not a required module for Airflow, but now any machine without it can't run any local unit tests! I put the traceback I'm seeing below.

The unit test file causing this import needs to be guarded so it only runs in if its dependencies are available (see `hive_operator.py` in the same tests directory for one example, or the explicit `skip` guards in the postgres/mysql tests that ensure they only run in the right travis environment).

```bash
======================================================================
ERROR: Failure: ImportError (No module named 'hive_metastore')
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/Users/jlowin/anaconda3/lib/python3.5/site-packages/nose/failure.py", line 39, in runTest
    raise self.exc_val.with_traceback(self.tb)
  File "/Users/jlowin/anaconda3/lib/python3.5/site-packages/nose/loader.py", line 418, in loadTestsFromName
    addr.filename, addr.module)
  File "/Users/jlowin/anaconda3/lib/python3.5/site-packages/nose/importer.py", line 47, in importFromPath
    return self.importFromDir(dir_path, fqname)
  File "/Users/jlowin/anaconda3/lib/python3.5/site-packages/nose/importer.py", line 94, in importFromDir
    mod = load_module(part_fqname, fh, filename, desc)
  File "/Users/jlowin/anaconda3/lib/python3.5/imp.py", line 244, in load_module
    return load_package(name, filename)
  File "/Users/jlowin/anaconda3/lib/python3.5/imp.py", line 216, in load_package
    return _load(spec)
  File "<frozen importlib._bootstrap>", line 693, in _load
  File "<frozen importlib._bootstrap>", line 673, in _load_unlocked
  File "<frozen importlib._bootstrap_external>", line 665, in exec_module
  File "<frozen importlib._bootstrap>", line 222, in _call_with_frames_removed
  File "/Users/jlowin/git/airflow/tests/__init__.py", line 24, in <module>
    from .operators import *
  File "/Users/jlowin/git/airflow/tests/operators/__init__.py", line 20, in <module>
    from .s3_to_hive_operator import *
  File "/Users/jlowin/git/airflow/tests/operators/s3_to_hive_operator.py", line 25, in <module>
    from airflow.operators.s3_to_hive_operator import S3ToHiveTransfer
  File "/Users/jlowin/git/airflow/airflow/operators/s3_to_hive_operator.py", line 27, in <module>
    from airflow.hooks.hive_hooks import HiveCliHook
  File "/Users/jlowin/git/airflow/airflow/hooks/hive_hooks.py", line 28, in <module>
    import hive_metastore
ImportError: No module named 'hive_metastore'
```

**Comment by krishnabhupatiraju** [comment]
@jlowin I added the line "from .s3_to_hive_operator import *" to __init__.py. Since I had hive_metastore installed on my venvI didnt realize it might cause an issue. Similarly travis didnt fail either. I will create a PR to fix this.

**Pull request closed without merging**